planner/core/expression_rewriter.go

```diff
@@ -329,24 +329,24 @@ func (er *expressionRewriter) buildSubquery(ctx context.Context, subq *ast.Subqu
 		}()
 	}
 	// Store the old value before we enter the subquery and reset they to default value.
-	oldRewriteHintCanTakeEffect := er.b.checkSemiJoinHint
-	er.b.checkSemiJoinHint = rewriteHintCanTakeEffect
-	oldHasHint := er.b.hasValidSemiJoinHint
-	er.b.hasValidSemiJoinHint = false
+	oldSubQCtx := er.b.subQueryCtx
+	er.b.subQueryCtx = subqueryCtx
+	oldHintFlags := er.b.subQueryHintFlags
+	er.b.subQueryHintFlags = 0
 	outerWindowSpecs := er.b.windowSpecs
 	defer func() {
 		er.b.windowSpecs = outerWindowSpecs
-		er.b.checkSemiJoinHint = oldRewriteHintCanTakeEffect
-		er.b.hasValidSemiJoinHint = oldHasHint
+		er.b.subQueryCtx = oldSubQCtx
+		er.b.subQueryHintFlags = oldHintFlags
 	}()
 
 	np, err = er.b.buildResultSetNode(ctx, subq.Query, false)
 	if err != nil {
-		return nil, false, err
+		return nil, 0, err
 	}
 	// Pop the handle map generated by the subquery.
 	er.b.handleHelper.popMap()
-	return np, er.b.hasValidSemiJoinHint, nil
+	return np, er.b.subQueryHintFlags, nil
```

Comment:
It seems we always return `0` for the return value `hintFlags uint64`? Because we set `er.b.subQueryHintFlags = 0` in line335 and return always execute before the defer function in line337-341.

Reply:
No. You can see the two examples to see golang's behavior:
https://go.dev/play/p/jg7rfs2Qc6f
https://go.dev/play/p/Cb7L2I03tvC

Reply:
It's indeed a bit confusing, I updated the code to make it more clear.